### PR TITLE
Draft: Add Metrics

### PR DESCRIPTION
**Pull request by Lookatator**
## Description

Add Main Metrics calculators. 
For the moment, we have a main abstract class `DiscreteArchiveMetricsCalculator`, which has 2 children:
- `GridMetricsCalculator` (using a standard MAP-Elites grid)
- `CVTMetricsCalculator` (using a CVT grid)

Each instance of those classes has a method `get_metrics` which return a dataclass object `DiscreteArchiveMetrics` (greatly inspired by the `ArchiveStats` of `pyribs`).

## TODO

- [ ] Complete Docstrings
- [ ] (future) Continuous QD scores calculators.
- [ ] (future) Add similar metrics calculators using `jax` in a separate folder (would make the metrics calculation faster and easier to vectorise).

## Questions

- [ ] There are 2 ways to calculate the QD-Score in the litterature: (1) adding an offset to all fitnesses and summing them, and (2) normalizing the fitnesses given a lower and an upper bound. For the moment, I implemented it using (2) only. Should I also add (1) ?

## Status

- [x] I have read the guidelines in CONTRIBUTING.md
- [x] I have formatted my code using `black`
- [ ] I have tested my code by running `pytest`
- [ ] I have added a one-line description of my change to the changelog in
      `HISTORY.md`
- [ ] This PR is ready to go

### CODE REVIEW

File: qdglue/metrics/grid_based_metrics.py [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+from qdglue.types import Feature, Fitness
+
+
+@dataclasses.dataclass
+class DiscreteArchiveMetrics:
+    """Holds statistics about an archive.
+
+    Attributes:
+        coverage: Proportion of cells in the archive that have an elite - always
+            in the range $[0,1]$.
```
Comment (btjanaka) [inline qdglue/metrics/grid_based_metrics.py:19]: I added the mkdocs arithmatex extension so writing with dollar signs works now. Careful with escaping backslashes though (probably just use a raw string if you need backslashes).

File: qdglue/metrics/grid_based_metrics.py [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+        qd_score: QD score, i.e. sum of objective values of all elites in the
```
Comment (btjanaka) [inline qdglue/metrics/grid_based_metrics.py]: I think we should normalize this QD score only by adding the min objective in order to be consistent with the original introduction of QD score in Pugh 2016.

I think we can add another field for the version of QD score with the upper bound, but I would prefer not to call it `norm_qd_score` since in pyribs we already refer to `norm_qd_score` as the QD score I mentioned above, divided by number of archive cells.

It might be helpful to have more info before deciding on a name. How does upper bounding the fitness affect the final QD score, and does the max value of the QD score simply become the number of cells in the archive? One name that comes to mind here is `bound_norm_qd_score`, although it is a bit lengthy.

File: qdglue/metrics/grid_based_metrics.py [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+                best_fitness_per_centroid_index[index_closest_centroid] = np.maximum(
+                    best_fitness_per_centroid_index[index_closest_centroid], fitness
+                )
+
+        best_fitnesses = np.asarray(list(best_fitness_per_centroid_index.values()))
+
+        return best_fitnesses
+
+
+def example_grid_based_metrics():
```
Comment (btjanaka) [inline qdglue/metrics/grid_based_metrics.py]: Move this to an example file or (even better) a test?

File: qdglue/metrics/grid_based_metrics.py [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+            fitness_bounds before computing the QD score.
+        max_fitness: Maximum objective value of the elites in the archive.
+        ccdf: Complementary Cumulative Distribution of Fitness from Vassiliades
+            et al. (2018)
+    """
+
+    coverage: float
+    qd_score: float
+    max_fitness: float
+    ccdf: np.ndarray
```
Comment (btjanaka) [inline qdglue/metrics/grid_based_metrics.py:37]: It seems the CCDF depends on what the edges of the bins are. I think this class should contain some way to reconstruct the bin edges -- for instance, it should have the edges themselves, or the fitness lower/upper bound with num_bins.

File: qdglue/metrics/grid_based_metrics.py [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+from typing import Dict, Tuple
+
+import gymnasium
+import numpy as np
+
+from qdglue.types import Feature, Fitness
+
+
+@dataclasses.dataclass
+class DiscreteArchiveMetrics:
```
Comment (btjanaka) [inline qdglue/metrics/grid_based_metrics.py:14]: I haven't had to time to think this through, but I think we should generally aim to make this class "self-contained" -- i.e. if all I get is this class, I shouldn't have to look back at my QD algorithm to figure out what these metrics mean. This way, I can easily pass this object around and someone else can interpret what the metrics mean.

File: qdglue/metrics/grid_based_metrics.py [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+            in the range $[0,1]$.
+        qd_score: QD score, i.e. sum of objective values of all elites in the
+            archive. All the fitnesses are normalized to $[0,1]$ using the
+            fitness_bounds before computing the QD score.
+        max_fitness: Maximum objective value of the elites in the archive.
+        ccdf: Complementary Cumulative Distribution of Fitness from Vassiliades
+            et al. (2018)
+    """
+
+    coverage: float
```
Comment (btjanaka) [inline qdglue/metrics/grid_based_metrics.py:33]: Going with the theme of self-containment, I think we should add some details on the archive resolution to this class.

File: qdglue/metrics/grid_based_metrics.py [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+            qd_score=qd_score,
+            max_fitness=max_fitness,
+            ccdf=ccdf,
+        )
+
+    def _compute_ccdf(
+        self,
+        best_fitnesses: Fitness,
+    ) -> np.ndarray:
+        fitnesses_intermediate = np.linspace(
```
Comment (btjanaka) [inline qdglue/metrics/grid_based_metrics.py]: My feeling is that `np.histogram` will do what you're trying to do here, but I can't remember how. Can you take a look? I think it will reduce this function to roughly one function call.

**Review by btjanaka — COMMENTED** [review]
Hi @Lookatator, great work! I haven't looked too much into the Calculator classes yet, but I left some comments on the metrics dataclass.

File: qdglue/metrics/grid_based_metrics.py [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+            et al. (2018)
+    """
+
+    coverage: float
+    qd_score: float
+    max_fitness: float
+    ccdf: np.ndarray
+
+
+class DiscreteArchiveMetricsCalculator(ABC):
```
Comment (btjanaka) [inline qdglue/metrics/grid_based_metrics.py:52]: The first thing I think about when I hear about metrics is a function, not really an object that I have to instantiate. Given that these classes are all essentially executing one function (get_metrics), what do you think of refactoring all of these into functions and helper functions? Or at least, maybe `get_metrics` should become `__call__`.

On the other hand, there are precedents for "calculator" functions that are also classes, e.g. loss functions in PyTorch https://pytorch.org/docs/stable/nn.html#loss-functions -- note that these also use `__call__`.

**Review by btjanaka — COMMENTED** [review]
More comments

File: qdglue/metrics/grid_based_metrics.py (hunk 1 above)
Comment (Lookatator) [reply qdglue/metrics/grid_based_metrics.py:19]: Thank you very much!

File: qdglue/metrics/grid_based_metrics.py (hunk 3 above)
Comment (Lookatator) [reply qdglue/metrics/grid_based_metrics.py]: moved to a test :)

File: qdglue/metrics/grid_based_metrics.py (hunk 2 above)
Comment (Lookatator) [reply qdglue/metrics/grid_based_metrics.py]: agree we should have a new field this "normalized" QD-Score, as it has been widely used in the community (e.g. in the QD Framework paper from Cully 2018, and in the DQD paper). Also it is more interpretable than the old one from Pugh 2016. 

- I added a QD score that follows the one of Pugh 2016 and called it `qd_score_original`.
- I like `qd_score_bound_norm` for the "normalized" QD-Score, and I don't mind if it's lengthy, especially if it describes well how it is calculated ^^

File: qdglue/metrics/grid_based_metrics.py (hunk 4 above)
Comment (Lookatator) [reply qdglue/metrics/grid_based_metrics.py:37]: I had those infos in the beginning, and removed them as I thought you would like the class to only contain the calculated metrics. So I added a `discrete_archive_metrics_calculator` attribute (a bit ugly structure, I know, but that's working super well and it's super convenient).

Also added a `ccdf_fitness_bins` property that refinds the fitness bins using `discrete_archive_metrics_calculator`.

File: qdglue/metrics/grid_based_metrics.py (hunk 5 above)
Comment (Lookatator) [reply qdglue/metrics/grid_based_metrics.py:14]: As said above, I was hesitating about that. And I do not have any strong opinion.

I just added a `discrete_archive_metrics_calculator` attribute to the metrics dataclass, so that we can find all the info about the metrics calculator directly in the metrics object.

File: qdglue/metrics/grid_based_metrics.py (hunk 6 above)
Comment (Lookatator) [reply qdglue/metrics/grid_based_metrics.py:33]: Addressed in the comments above :D

File: qdglue/metrics/grid_based_metrics.py (hunk 7 above)
Comment (Lookatator) [reply qdglue/metrics/grid_based_metrics.py]: Nice! `np.histogram` with an inverted `cumsum` does the job! Thank you so much! ^^

File: qdglue/metrics/grid_based_metrics.py (hunk 8 above)
Comment (Lookatator) [reply qdglue/metrics/grid_based_metrics.py:52]: I thought about several structures. The reasons why I chose objects are:
- I like the idea that someone creates an object of type `GridMetricsCalculator`, and gives the parameters (e.g. grid resolution, fitness bounds, feature space) only once, and then use this single object to calculate the metrics for all variants and all replications
- If I used functions and helper functions:
    - the number of arguments could be quite long, for the main function and for the helper functions...
    - if the user wants to fix some parameters (e.g. grid resolution, feature space...) he would have to know and ideally use `partial`. And I don't like `partial`. One of the reasons is that many people do not how it works, and I find it quite ugly. So I try to avoid it where I can.

Comment (Lookatator) [reply qdglue/metrics/grid_based_metrics.py:52]: And I don't like using only `__call__` for code that is intended for a wide audience, as many people do not know it.

As an in-between solution, I created a method `__call__` that directly calls `get_metrics` (as done in PyTorch)

File: qdglue/metrics/grid_based_metrics.py (hunk 4 above)
Comment (btjanaka) [reply qdglue/metrics/grid_based_metrics.py:37]: Another property that would be good to have is that we can save this class and reload it across versions of Python and even versions of the library. For instance, it would be nice if we could pickle this class (or at least save it as JSON) in version 0.1.0, and then in version 0.5.0 we're still able to open it. Think about someone who evaluates their QD algorithm and stores the metrics in their repo. Then someone else comes along a couple years later, runs their QD algorithm, and is able to easily compare metrics.

I'm not yet sure what we need to do to achieve this property -- it's very difficult due to the time factor and we should probably look around. However, one thing that would make this more difficult is storing a class type like `DiscreteArchiveMetricsCalculator`, which is likely to break across versions and thus not work with pickle. I would prefer that we store info from the calculator class directly in this class; e.g., we can put the fitness bounds and ccdf fitness bins.

Regarding ccdf_fitness_bins, I think we can just store this directly on the class instead of creating the property. Keep in mind that:

- A property method gets re-evaluated every time you call it.
- The ccdf_fitness_bins are not necessarily evenly spaced, i.e., they may vary based on the calculator.
- This is a dataclass. I think we should try to make sure it strictly stores data, with no methods at all. Methods are also much more prone than data fields to breaking across versions.

File: qdglue/metrics/grid_based_metrics.py (hunk 5 above)
Comment (btjanaka) [reply qdglue/metrics/grid_based_metrics.py:14]: See other comment below.

File: qdglue/metrics/grid_based_metrics.py (hunk 8 above)
Comment (btjanaka) [reply qdglue/metrics/grid_based_metrics.py:52]: These are fair points; using `__call__` also opens up the opportunity for us to put wrappers around `get_metrics` like input validation and output validation in the future if the need should arise. Let's keep it as is, but if we add more classes, let's make sure it remains a one-layer hierarchy for simplicity.

File: qdglue/metrics/grid_based_metrics.py [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+        )
+
+
+class DiscreteArchiveMetricsCalculator(ABC):
+    def __init__(
+        self,
+        fitness_bounds: gymnasium.spaces.Box,
+        num_points_ccdf: int = 100,
+    ):
+        assert np.all(np.isfinite(fitness_bounds.low)) and np.all(
```
Comment (btjanaka) [inline qdglue/metrics/grid_based_metrics.py:58]: Unsure if we should restrict our fitness to be within certain bounds. There are a host of problems where we don't know the upper or even the lower bound, e.g., many QD-RL locomotion domains.

Comment (btjanaka) [inline qdglue/metrics/grid_based_metrics.py:58]: We should raise `ValueError` here -- assertions are too broad; something like `if X: raise ValueError()`. I generally follow the exception descriptions listed here: https://docs.python.org/3/library/exceptions.html#concrete-exceptions

File: qdglue/metrics/grid_based_metrics.py [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+    def _check_validity_features(self, features: Feature) -> None:
+        """Check validity of features."""
+
+    def _check_validity_fitnesses(self, fitnesses):
+        assert np.all(np.min(fitnesses, axis=0) >= self.fitness_bounds.low) and np.all(
+            np.max(fitnesses, axis=0) <= self.fitness_bounds.high
+        ), "Fitnesses must be within the fitness bounds"
+
+    @abstractmethod
+    def _extract_best_fitnesses_per_cell(
```
Comment (btjanaka) [inline qdglue/metrics/grid_based_metrics.py:143]: It seems you're using inheritance here primarily as a way to reuse code, which means that child classes are left to implement a seemingly random set of abstract methods (compare: just implementing forward() in pytorch).

The design depends a lot on our goal here. If we just want to make calculators that the user can use, then this works fine. However, if we want the user to be able to implement their own calculators, we'll have to think a lot more about how these classes should be designed. We may want to think of how we can use composition instead of inheritance here, for instance.

File: qdglue/metrics/grid_based_metrics.py (hunk 9 above)
Comment (Lookatator) [reply qdglue/metrics/grid_based_metrics.py:58]: Agreed! 
[TODO] if we go on with the current structure

Comment (Lookatator) [reply qdglue/metrics/grid_based_metrics.py:58]: Agreed that the _task_ itself has no lower and upper bounds.
However, some _metrics_ (such as `QD-score`) require a minimal fitness. 
Such minimal fitness is sometimes calculated from all the results from all variants on a task.

[TO BE DISCUSSED]

File: qdglue/metrics/grid_based_metrics.py (hunk 4 above)
Comment (Lookatator) [reply qdglue/metrics/grid_based_metrics.py:37]: I agree with most of your comments.
In particular, the `ccdf_fitness_bins` property was not an amazing idea.

[TO BE DISCUSSED]
